neuraxle/higher_order_steps.py

```diff
@@ -0,0 +1,117 @@
+
+..
+    Copyright 2019, Neuraxio Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+"""
+from abc import ABC
+
+from neuraxle.base import BaseTransformer, ExecutionContext, MetaStep, ForceHandleMixin
+from neuraxle.data_container import DataContainer
+
+
+class BaseHigherOrderStep(ForceHandleMixin, MetaStep, ABC):
+    """
+    Base class to represent an higher order step.
+    A higher-order step is just a function that takes an existing step and returns another step that wraps it.
+
+    An higher order step:
+        - cannot be saved
+        - cannot modify the data container
+        - cannot push himself in the execution context
+
+    .. seealso::
+        :class:`~neuraxle.base.MetaStep`
+    """
+    def __init__(self, step: BaseTransformer):
+        MetaStep.__init__(self, wrapped=step)
+        ForceHandleMixin.__init__(self)
+
+    def save(self, context: ExecutionContext, full_dump=False):
+        """
+        Don't save the higher order step. Save the wrapped step only.
+
+        :param context: context to save.
+        :param full_dump: save full dump or not
+        :return: saved wrapped step
+        """
+        if len(context) > 0 and isinstance(context.parents[-1], BaseHigherOrderStep):
+            context.pop()
+        return self.wrapped.save(context, full_dump=full_dump)
+
+    def _will_process(self, data_container: DataContainer, context: ExecutionContext) -> (DataContainer, ExecutionContext):
+        return data_container, context
+
+    def _will_fit(self, data_container: DataContainer, context: ExecutionContext) -> (DataContainer, ExecutionContext):
+        return data_container, context
+
+    def _will_transform_data_container(self, data_container: DataContainer, context: ExecutionContext) -> (DataContainer, ExecutionContext):
+        return data_container, context
+
+    def _will_fit_transform(self, data_container: DataContainer, context: ExecutionContext):
+        return data_container, context
+
+    def _did_fit_transform(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
+        return data_container
+
+    def _did_process(self, data_container: DataContainer, context: ExecutionContext):
+        return data_container
+
+    def _did_transform(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
+        return data_container
+
+
+def WithContext(step: BaseTransformer, context: ExecutionContext):
```

Comment:
Two choices here: 
- A) move this method to `BaseStep.with_context`, **just like .tosklearn()** as I indicated in the previous review comment.
- B) otherwise please delete `def WithContext` and just use `class WithContext` with a constructor that accepts the context and sets it in self because the method does nothing more here (except than hiding the context at an other place than self which will be harmful to the new saver). 

```suggestion
def with_context(self, step: BaseTransformer, context: ExecutionContext):
```

Reply:
(oh, side note: nice docstring. Please put it in the `class StepWithContext` if your choice is to delete the new method, otherwise the docstring is super cool).

Reply:
@guillaume-chevalier those changes CANNOT be done because of the references. I've tried moving things around for an hour, and it doesn't make sense. base.py is becoming way to big. the new file was fine imo. we should use a wrapper directly as well.

Reply:
awaiting for your response to continue @guillaume-chevalier